### PR TITLE
Change the GCA algorithm

### PR DESCRIPTION
**Pull request by dzajkowski**
## Overview
<sup>_What this PR does, and why it's needed_</sup>
Introduces an algorithm that does not need to walk the dag from genesis.

### JIRA ticket:
<sup>_Create it if there isn't one already._</sup>
RCHAIN-3002

### Notes
<sup>_Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else._</sup>

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

File: casper/src/main/scala/coop/rchain/casper/util/DagOperations.scala [hunk 1]
```diff
@@ -97,49 +92,31 @@ object DagOperations {
 
   /**
     * Conceptually, the GCA is the first point at which the histories of b1 and b2 diverge.
-    * We compute by finding the first block from genesis for which there exists a child of that block which is an ancestor of b1 or b2 but not both.
-    *
-    * TODO: Implement a GCA that doesn't require the genesis block (see https://rchain.atlassian.net/browse/RCHAIN-3002)
-    * TODO: Remove usage of BlockStore by just using BlockDagRepresentation (see https://rchain.atlassian.net/browse/RCHAIN-3003)
+    * We compute by finding the first block that is the "oldest" (has highest blocknum) common block.
```
Comment (KentShikama) [inline casper/src/main/scala/coop/rchain/casper/util/DagOperations.scala]: The genesis block which has a block number of 0 is the "oldest"?

File: casper/src/main/scala/coop/rchain/casper/util/DagOperations.scala [hunk 2]
```diff
@@ -97,49 +92,31 @@ object DagOperations {
...
-                )
-      } yield gca.get
+      val start: SortedSet[BlockWithDepth] = SortedSet
+        .empty[BlockWithDepth] + ((b1.blockNum, b1)) + ((b2.blockNum, b2))
+      Monad[F]
+        .iterateWhileM(start)(s => {
+          val (head, tail) = (s.head._2, s.tail)
+          getParents(head).map(nb => tail ++ nb.map(v => (v.blockNum, v)))
+        })(_.size != 1)
+        .map(v => v.head._2)
```
Comment (KentShikama) [inline casper/src/main/scala/coop/rchain/casper/util/DagOperations.scala]: I haven't really tried hard yet but I would appreciate if you could make this more readable somehow.

File: casper/src/main/scala/coop/rchain/casper/util/DagOperations.scala [hunk 3]
```diff
@@ -96,50 +91,32 @@ object DagOperations {
...
-                                   for {
-                                     c <- ProtoUtil.unsafeGetBlock[F](hash)
-                                   } yield b1Ancestors(c) ^ b2Ancestors(c)
-                               )
-                    } yield result
-                )
-      } yield gca.get
+      val start = SortedSet.empty[BlockWithDepth] + ((b1.blockNum, b1)) + ((b2.blockNum, b2))
+      val result = Monad[F].iterateWhileM(start)(extractParentsFromHighestNumBlock)(_.size != 1).map(v => v.head)
+      result.map(_._2)
```
Comment (odeac) [inline casper/src/main/scala/coop/rchain/casper/util/DagOperations.scala]: the two consecutive maps and the result intermediate value can be eliminated without loss of readability.

Also it's more readable if written with one function call per line, like this:
```scala
Monad[F].iterateWhileM(start)(extractParentsFromHighestNumBlock)(_.size != 1)
       .map(v => v.head._2)
```

File: casper/src/main/scala/coop/rchain/casper/util/DagOperations.scala [hunk 4]
```diff
@@ -96,50 +91,32 @@ object DagOperations {
...
+    type BlockWithDepth = (Long, BlockMetadata)
+    implicit val blockNumDecreasing: Ordering[BlockWithDepth] =
+      (l: BlockWithDepth, r: BlockWithDepth) => r._1.compare(l._1)
+
+    def getParents(p: BlockMetadata): F[Set[BlockMetadata]] =
+      p.parents.map(v => dag.lookup(v)).sequence.map(_.toSet.flatten)
+
+    def extractParentsFromHighestNumBlock(s: SortedSet[BlockWithDepth]): F[SortedSet[BlockWithDepth]] = {
+      val (head, tail) = (s.head._2, s.tail)
+      getParents(head).map(nb => tail ++ nb.map(v => (v.blockNum, v)))
```
Comment (odeac) [inline casper/src/main/scala/coop/rchain/casper/util/DagOperations.scala]: Is difficult to understand what `nb => tail ++ ...` does. Please extract it as a inner named function for improved readability.

File: casper/src/main/scala/coop/rchain/casper/util/DagOperations.scala [hunk 5]
```diff
@@ -96,50 +91,32 @@ object DagOperations {
...
+      b2: BlockMetadata,
       dag: BlockDagRepresentation[F]
-  ): F[BlockMessage] =
+  ): F[BlockMetadata] = {
+    type BlockWithDepth = (Long, BlockMetadata)
+    implicit val blockNumDecreasing: Ordering[BlockWithDepth] =
+      (l: BlockWithDepth, r: BlockWithDepth) => r._1.compare(l._1)
+
+    def getParents(p: BlockMetadata): F[Set[BlockMetadata]] =
+      p.parents.map(v => dag.lookup(v)).sequence.map(_.toSet.flatten)
```
Comment (odeac) [inline casper/src/main/scala/coop/rchain/casper/util/DagOperations.scala]:
```scala
x.map(f).sequence
``` 
is the same as 
```scala 
x.traverse(f)
```
So this code can be written:
```scala
p.parents
  .traverse(v => dag.lookup(v))
  .map(_.toSet.flatten)
```

File: casper/src/test/scala/coop/rchain/casper/util/DagOperationsTest.scala [hunk 6]
```diff
@@ -40,20 +44,36 @@ class DagOperationsTest
         for {
           genesis <- createBlock[Task](Seq.empty)
           b1      <- createBlock[Task](Seq(genesis.blockHash))
+          b1m     = BlockMetadata.fromBlock(b1)
```
Comment (odeac) [inline casper/src/test/scala/coop/rchain/casper/util/DagOperationsTest.scala]: For less repetition and improved readability you could use a helper function which returns a case class containing the block and its metadata. 

I'm thinking something like this:

```scala
final case class BlockWithMeta(val block, val metadata)

def createBlockWithMeta(...) = {...}

for {
 b1 <- createBlockWithMeta(...)
 b2 <- createBlockWithMeta(b1.block.blockHash)
 ...
_ <- DagOperations.lowestCommonAncestorF[Task](b1.metadata, b5.metadata, dag) shouldBeF b1.metadata
}
```

File: casper/src/main/scala/coop/rchain/casper/Estimator.scala [hunk 7]
```diff
@@ -50,23 +51,30 @@ object Estimator {
       sortedChildren      = maybeSortedChildren.flatten.toVector
     } yield sortedChildren
 
-  private def calculateGca[F[_]: Monad: BlockStore](
+  private def calculateGca[F[_]: Monad](
```
Comment (KentShikama) [inline casper/src/main/scala/coop/rchain/casper/Estimator.scala]: Now Lca

**Pull request merged**